libp2p/network/url_conversion.go

```diff
@@ -0,0 +1,107 @@
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/multiformats/go-multiaddr"
+)
+
+const (
+	// anyScheme is a string which fills the "scheme" degree of freedom for
+	// a URL. Used to pad URLs without schemes such that they can be parsed
+	// via stdlib URL parser.
+	anyScheme = "scheme://"
+)
+
+// Libp2pMultiaddrFromServiceUrl transforms a URL into its libp2p multiaddr equivalent.
+// (see: https://github.com/libp2p/specs/blob/master/addressing/README.md#multiaddr-basics)
+func Libp2pMultiaddrFromServiceUrl(serviceUrl string) (multiaddr.Multiaddr, error) {
+	peerUrl, err := url.Parse(anyScheme + serviceUrl)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"parsing peer service URL: %s: %w",
+			serviceUrl,
+			err,
+		)
+	}
+
+	var (
+		// TECHDEBT: assuming TCP; remote peer's transport type must be knowable!
+		// (ubiquitously switching to multiaddr, instead of a URL, would resolve this)
+		peerTransportStr = "tcp"
+		peerIPVersionStr = "ip4"
+	)
+
+	// TECHDEBT: there's probably a more conventional way to do this.
+	// Check if we're dealing with IPv4 or IPv6
+	if strings.Count(peerUrl.Hostname(), ":") > 0 {
+		peerIPVersionStr = "ip6"
+	}
+
+	/* CONSIDER: using a `/dns<4 or 6>/<hostname>` multiaddr instead of resolving here.
+	 * I attempted using `/dns4/.../tcp/...` and go this error:
+	 * > failed to listen on any addresses: [can only dial TCP over IPv4 or IPv6]
+	 *
+	 * TECHDEBT: is there a way for us to effectively prefer IPv6 responses when resolving DNS?
+	 * TECHDEBT: resolving DNS this way has limitations:
+	 * > The address parameter can use a host name, but this is not recommended,
+	 * > because it will return at most one of the host name's IP addresses.
```

Comment:
Which one specifically? Just a random one?

Reply:
I would have to read a bit through stdlib to be able to answer that, the lines prefixed with `> `  are copy/pasted from the doc page on the line below.